types/mongodb/index.d.ts

```diff
@@ -58,13 +59,136 @@ export class MongoClient extends EventEmitter {
+ */
+export interface ClientSession extends EventEmitter {
+    /** The server id associated with this session */
+    id: any;
+    /**
+     * Aborts the currently active transaction in this session. 
+     * @param {MongoCallback<void>} [cb] Optional callback for completion of this operation
+     */
+    abortTransaction(cb?: MongoCallback<void>): Promise<void>;
+    /**
+     * Advances the operationTime for a ClientSession.
+     * @param {mongodb.Timestamp} operationTime
+     */
+    advanceOperationTime(operamtionTime: Timestamp): void;
+    /**
+     * Commits the currently active transaction in this session. 
+     * @param {MongoCallback<void>} [cb] Optional callback for completion of this operation
+     */
+    commitTransaction(cb?: MongoCallback<void>): Promise<void>;
+
+    /**
+     * Ends this session on the server
+     * @param {MongoCallback<void>} [cb] Optional callback for completion of this operation
+     */
+    endSession(cb?: MongoCallback<void>): void;
+    /**
+     * Ends this session on the server
+     * @param {*} [options] Optional settings. Currently reserved for future use
+     * @param {MongoCallback<void>} [cb] Optional callback for completion of this operation
+     */
+    endSession(options: any, cb?: MongoCallback<void>): void
 
-declare class ClientSession extends EventEmitter {
-    endSession(callback?: MongoCallback<void>): void;
-    endSession(options: any, callback?: MongoCallback<void>): void;
+    /**
+     * Used to determine if this session equals another
+     *
+     * @param {ClientSession} session A class representing a client session on the server
+     * @returns {boolean} if the sessions are equal 
+     */
     equals(session: ClientSession): boolean;
+
+    /** Increment the transaction number on the internal ServerSession */
+    incrementTransactionNumber(): void;
+
+    /**
+     * @returns {boolean} this session is currently in a transaction or not 
+     */
+    inTransaction(): boolean;
+
+    /**
+     * Starts a new transaction with the given options.
+     * @param {TransactionOptions} options 
+     * @memberof ClientSession
+     */
+    startTransaction(options?: TransactionOptions): void;
+
+}
+
+// http://mongodb.github.io/node-mongodb-native/3.1/api/global.html#ReadConcern
+type ReadConcernLevel = 'local' | 'available' | 'majority' | 'linearizable' | 'snapshot';
+
+/**
+ * The MongoDB ReadConcern, which allows for control of the consistency and isolation properties
+ * of the data read from replica sets and replica set shards.
+ * http://mongodb.github.io/node-mongodb-native/3.1/api/global.html#ReadConcern
+ */
+export interface ReadConcern {
+    level: ReadConcernLevel;
+}
+
+/**
+ * A MongoDB WriteConcern, which describes the level of acknowledgement
+ * requested from MongoDB for write operations.
+ * http://mongodb.github.io/node-mongodb-native/3.1/api/global.html#WriteConcern
+ */
+interface WriteConcern {
```

Comment:
Can you also update `CommonOptions ` at line 455 and `GridFSBucketWriteStreamOptions ` at line 1626 to make them extend this interface? Just to avoid duplicating the interfaces

Reply:
updated the interface 9c3770f940c74b04b66a71156d52f42fbc6e037a
unfortunately the options of  GridFSBucketWriteStreamOptions differ from WriteConcern